velox/functions/sparksql/aggregates/FirstLastAggregate.cpp

```diff
@@ -105,17 +90,54 @@ class FirstLastAggregateBase
 
   void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
       override {
-    extractValues(groups, numGroups, result);
+    auto rowVector = (*result)->as<RowVector>();
+    VELOX_CHECK_EQ(
+        rowVector->childrenSize(),
+        2,
+        "intermediate results must have 2 children");
+
+    auto ignoreNullVector = rowVector->childAt(1)->asFlatVector<bool>();
+    rowVector->resize(numGroups);
+    ignoreNullVector->resize(numGroups);
+
+    extractValues(groups, numGroups, &(rowVector->childAt(0)));
+    for (auto i = 0; i < numGroups; i++) {
+      if (Aggregate::isNull(groups[i])) {
+        rowVector->setNull(i, true);
+      }
+    }
   }
 
   void destroy(folly::Range<char**> groups) override {
     if constexpr (!numeric) {
       for (auto group : groups) {
-        auto accumulator = exec::Aggregate::value<TAccumulator>(group);
-        accumulator->value().destroy(exec::Aggregate::allocator_);
+        auto accumulator = Aggregate::value<TAccumulator>(group);
+        // If ignoreNull is true and groups are all null, accumulator will not
+        // set.
+        if (accumulator->has_value()) {
+          accumulator->value().destroy(Aggregate::allocator_);
+        }
       }
     }
   }
+
+ protected:
+  DecodedVector decodedValue_;
```

Comment:
please, move member variables at the end of the class, after decodeIntermediateRows method